src/document.rs

```diff
@@ -200,6 +200,71 @@ pub(crate) fn extract_value(any: &PyAny) -> PyResult<Value> {
     Err(to_pyerr(format!("Value unsupported {any:?}")))
 }
 
+pub(crate) fn extract_value_for_type(
+    any: &PyAny,
+    tv_type: tv::schema::Type,
+    field_name: &str,
+) -> PyResult<Value> {
+    // Helper function to create `PyErr`s returned by this function.
+    fn to_pyerr_for_type<'a, E: std::error::Error>(
+        type_name: &'a str,
+        field_name: &'a str,
+        any: &'a PyAny,
+    ) -> impl Fn(E) -> PyErr + 'a {
+        move |_| {
+            to_pyerr(format!(
+                "Expected {} type for field {}, got {:?}",
+                type_name, field_name, any
+            ))
+        }
+    }
+
+    let value = match tv_type {
+        tv::schema::Type::Str => Value::Str(
+            any.extract::<String>()
+                .map_err(to_pyerr_for_type("Str", field_name, any))?,
+        ),
+        tv::schema::Type::U64 => Value::U64(
+            any.extract::<u64>()
+                .map_err(to_pyerr_for_type("U64", field_name, any))?,
+        ),
+        tv::schema::Type::I64 => Value::I64(
+            any.extract::<i64>()
+                .map_err(to_pyerr_for_type("I64", field_name, any))?,
+        ),
+        tv::schema::Type::F64 => Value::F64(
+            any.extract::<f64>()
+                .map_err(to_pyerr_for_type("F64", field_name, any))?,
+        ),
+        tv::schema::Type::Date => {
+            let py_datetime = any
+                .downcast::<PyDateTime>()
+                .map_err(to_pyerr_for_type("DateTime", field_name, any))?;
+            let datetime = Utc
+                .with_ymd_and_hms(
+                    py_datetime.get_year(),
+                    py_datetime.get_month().into(),
+                    py_datetime.get_day().into(),
+                    py_datetime.get_hour().into(),
+                    py_datetime.get_minute().into(),
+                    py_datetime.get_second().into(),
+                )
+                .single()
+                .unwrap();
+
+            Value::Date(tv::DateTime::from_timestamp_secs(datetime.timestamp()))
+        }
+        tv::schema::Type::Facet => Value::Facet(
+            any.extract::<Facet>()
+                .map_err(to_pyerr_for_type("Facet", field_name, any))?
+                .inner,
+        ),
+        _ => return Err(to_pyerr(format!("Value unsupported {:?}", any))),
+    };
+
+    Ok(value)
+}
+
 fn extract_value_single_or_list(any: &PyAny) -> PyResult<Vec<Value>> {
     if let Ok(values) = any.downcast::<PyList>() {
```

Comment:
Not related to this PR, but I think this would benefit from using `.extract::<Vec<Value>>()` or at least `.extract::<Vec<&PyAny>>()` instead of `.downcast::<PyList>()` to handle any sequence and not just lists.

Reply:
Would this needlessly create a temporary `Vec`? I'm transforming/mapping the elements inside, so maybe it's not worth collecting until the very end?

Reply:
> Would this needlessly create a temporary Vec? I'm transforming/mapping the elements inside, so maybe it's not worth collecting until the very end?

Yes, that would be the cost for supporting arbitrary sequences (like 1d NumPy arrays) instead of just lists. It is a trade-off and since it isn't really part of this PR it is probably best left to do elsewhere if done at all.

(Alternatively, this could use `any.iter()` to just iterate of the sequence, but `.iter()` also works for e.g. strings which would need to be checked before doing this.)